File: Algorithms/538. Convert BST to Greater Tree/javascript/solution.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
var convertBST = function (root) {
    let total = 0,
        node = root;
    while (node) {
        if (!node.right) {
            total += node.val;
            node.val = total;
            node = node.left;
        } else {
            const succ = getSuccessor(node);
            if (succ.left) {
                succ.left = null;
                total += node.val;
                node.val = total;
                node = node.left;
            } else {
                succ.left = node;
                node = node.right;
            }
        }
    }

    return root;
};

const getSuccessor = (node) => {
    let succ = node.right;
    while (succ.left && succ.left !== node) {
        succ = succ.left;
    }
    return succ;
};

File: Algorithms/538. Convert BST to Greater Tree/golang/solution.go
func getSuccessor(node *TreeNode) *TreeNode {
	succ := node.Right
	for succ.Left != nil && succ.Left != node {
		succ = succ.Left
	}
	return succ
}

func convertBST(root *TreeNode) *TreeNode {
	total := 0
	node := root
	for node != nil {
		if node.Right == nil {
			total += node.Val
			node.Val = total
			node = node.Left
		} else {
			succ := getSuccessor(node)
			if succ.Left == nil {
				succ.Left = node
				node = node.Right
			} else {
				succ.Left = nil
				total += node.Val
				node.Val = total
				node = node.Left
			}
		}
	}
	return root
}
